aten/src/ATen/native/cuda/SoftMax.cu

```diff
@@ -440,6 +459,58 @@ ilpReduce(index_t shift,
   return threadVal;
 }
 
+// Uses vectorized LD/ST instructions and computes the reduce using
+// Max and Add simultaneously in a single pass.
+template <int ILP, typename T, typename AccumT, typename index_t=int>
+__device__ __forceinline__ void
+ilpReduceMaxAndSum(index_t shift,
+          const T* data,
+          index_t size,
+          AccumT &threadMax,
+          AccumT &threadExp)
+{
+  using LoadT = at::native::memory::aligned_vector<T, ILP>;
+  index_t offset = threadIdx.x;
+
+  MaxFloat<T, AccumT> redFuncMax;
+  Add<AccumT> redFuncAdd;
+
+  // shift and do 1
+  if(shift > 0){
+    data -= shift;
+    size += shift;
+    if(threadIdx.x >= shift){
+      T val = data[offset];
+      threadMax = redFuncMax(threadMax, val);
+      threadExp = redFuncAdd(threadExp, std::exp(val));
+    }
+    size -= blockDim.x;
+    data += blockDim.x;
+  }
+  index_t last = size % (ILP * blockDim.x);
+
+  T v[ILP];
+  LoadT* value = reinterpret_cast<LoadT*>(&v);
```

Comment:
Might not be a 100% correct code (but imo this is how what you are trying to do should be expressed)
But also, why do you need `LoadT` at all, is it because copying from vector to vector is faster?
```suggestion
  static_assert (sizeof(T)*ILP == sizeof(template LoadT));
  union {
     T v[ILP];
     LoadT value;
   } x;
```